proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,671 @@
+indexed arrow.
+
+- Multiplicities are a datatype:
+
+  ::
+
+    data Multiplicity
+      = One
+      | Omega
+      | Multiplicity :* Multiplicity
+      | Multiplicity :+ Multiplicity
+
+  In the following, for conciseness, we write ``1`` for ``One`` and
+  ``U`` (ASCII) or ``ω`` (Unicode) for ``Omega``. Note: unification
+  of multiplicities will be performed up to the semiring laws.
+- The multiplicity annotated arrow, for polymorphism, is written ``a:p
+  -> b`` (where ``a`` and ``b`` are types and ``p`` is a
+  multiplicity). To avoid introducing a new notion of "mixfix"
+  operators, we introduce a familly of (infix) type constructors:
+  ``(:p ->)`` for each multiplicity ``p``.
+
+The linear and unrestricted arrows are aliases:
+
+- ``(->)`` is an alias for ``(:'U ->)``
+- ``(->.)`` (ASCII syntax) and ``(⊸)`` (Unicode syntax) are aliases
+  for ``(:'1 ->)``
+
+Constructors & pattern-matching
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Constructors of data types defined with the Haskell'98 syntax
+
+::
+
+  data Foo
+    = Bar A B
+    | Baz C
+
+have linear function types, that is ``Bar :: A ->. B ->. Foo``. This
+is true in every module, including those without ``-XLinearTypes``
+turned on. This implies that most types in ``base`` (``Maybe``,
+``[]``, etc…) have linear constructors. We also make primivitive
+tuples ``(,)`` have linear constructors.
+
+With the GADT syntax, multiplicity of the arrows is honored:
+
+::
+
+  data Foo2 where
+    Bar2 :: A ->. B -> C
+
+then ``Bar2 :: A ->. B -> C``
+
+The definition of consuming a value in a data type exactly once must
+be refined to take the multiplicities of fields into account:
+
+- Consuming a value in a datatype exactly once means evaluating it to
+  head normal form and consuming its *linear* fields exactly once
+
+When pattern macthing a linear argument, linear fields are introduced
+as linear variables, and unrestricted fields as unrestricted
+variables:
+
+::
+
+  f :: Foo2 ->. A
+  f (Bar2 x y) = x  -- y is unrestricted, hence does not need to be consumed
+
+Base
+~~~~
+
+Because linear functions only strengthen the contract of unrestricted
+functions, a number of functions of ``base`` can get a more precise
+type. However, for pedagogical reason, to prevent linear types from
+interfering with newcomers' understanding the ``Prelude``, this
+proposal does not modify ``base``. Instead we will release a library
+exposing the stronger types for ``base`` functions.
+
+The only function which will need to change is ``($)`` because its
+typing rules is built in the type checker. Ignoring the details about
+levity and higher-rank polymorphism in the typing rule, the type
+``($)`` will be:
+
+::
+
+  ($) :: (a:p -> b) ⊸ a:p -> b
+
+The precise content of the library is out of scope of this proposal,
+but it will also contain convenient types to work with linear types
+such as:
+
+::
+
+   data Unrestricted a where
+     Unrestricted :: a -> Unrestricted a
+
+Formalism
```

Comment:
I'm delaying review of this section as I understand it is going to be rewritten.